pkg/pki/issue.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pki
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"hash/fnv"
+	"math/big"
+	"net"
+	"sort"
+	"strings"
+	"time"
+)
+
+var wellKnownCertificateTypes = map[string]string{
+	"ca":           "CA,KeyUsageCRLSign,KeyUsageCertSign",
+	"client":       "ExtKeyUsageClientAuth,KeyUsageDigitalSignature",
+	"clientServer": "ExtKeyUsageClientAuth,ExtKeyUsageServerAuth,KeyUsageDigitalSignature,KeyUsageKeyEncipherment",
+	"server":       "ExtKeyUsageServerAuth,KeyUsageDigitalSignature,KeyUsageKeyEncipherment",
+}
+
+type IssueCertRequest struct {
+	// Signer is the keypair to use to sign.
+	Signer string
+	// Type is the type of certificate i.e. CA, server, client etc.
+	Type string
+	// Subject is the certificate subject.
+	Subject pkix.Name
+	// AlternateNames is a list of alternative names for this certificate.
+	AlternateNames []string
+
+	// PrivateKey is the private key for this certificate. If nil, a new private key will be generated.
+	PrivateKey *PrivateKey
+	// MinValidDays is the lower bound on the certificate validity, in days. If specified, up to 30 days
+	// will be added so that certificate generated at the same time on different hosts will be unlikely to
+	// expire at the same time. The default is 10 years (without the up to 30 day skew).
+	MinValidDays int
+
+	// Serial is the certificate serial number. If nil, a random number will be generated.
+	Serial *big.Int
+}
+
+type Keystore interface {
+	// FindKeypair finds a cert & private key, returning nil where either is not found
+	// (if the certificate is found but not keypair, that is not an error: only the cert will be returned).
+	// This func returns a cert, private key and a bool.  The bool value is whether the keypair is stored
+	// in a legacy format. This bool is used by a keypair
+	// task to convert a Legacy Keypair to the new Keypair API format.
+	FindKeypair(name string) (*Certificate, *PrivateKey, bool, error)
+}
+
+func IssueCert(request *IssueCertRequest, keystore Keystore) (*Certificate, *PrivateKey, *Certificate, error) {
```

Comment:
Can this function get a comment? Mostly because it has two identical return types